service/common/src/main/java/org/apache/polaris/service/admin/PolarisAdminService.java

```diff
@@ -1650,15 +1664,42 @@ public List<GrantResource> listGrantsForCatalogRole(String catalogName, String c
    * @param entitiesMap map of entities
    * @param catalogId the id of the catalog of the entity we are looking for
    * @param id id of the entity we are looking for
+   * @param entityType
    * @return null if the entity does not exist
    */
   private @Nullable PolarisBaseEntity getOrLoadEntity(
-      @Nullable Map<Long, PolarisBaseEntity> entitiesMap, long catalogId, long id) {
+      @Nullable Map<Long, PolarisBaseEntity> entitiesMap,
+      long catalogId,
+      long id,
+      PolarisEntityType entityType) {
     return (entitiesMap == null)
-        ? metaStoreManager.loadEntity(getCurrentPolarisContext(), catalogId, id).getEntity()
+        ? metaStoreManager
+            .loadEntity(getCurrentPolarisContext(), catalogId, id, entityType)
+            .getEntity()
         : entitiesMap.get(id);
   }
 
+  private @Nullable PolarisBaseEntity getOrLoadEntityForGrant(
+      @Nullable Map<Long, PolarisBaseEntity> entitiesMap, PolarisGrantRecord record) {
+    if (entitiesMap != null) {
+      return entitiesMap.get(record.getSecurableId());
+    }
+
+    for (PolarisEntityType type : PolarisEntityType.values()) {
```

Comment:
It's kind of nice that the default behavior here will already let the persistence impls which happen not to use PolarisEntityType to short-circuit and return the entity on the first lookup, but perhaps we should just officially push-down the idea of "ANY"/"UNKNOWN" type to the lower layer and make it the persistence impl's responsibility to perform the for-loop.

We could allow `NULL_TYPE` to indicate the persistence layer should fan-out itself, or I guess we could add `ANY_TYPE` as `-1` the sme way `PolarisEntitySubType` has `ANY_SUBTYPE(-1, null)`.

That would also enable impls which group some set of types into one place and other sets of entities into another place to only have to fan-out over the number of *groupings* rather than let the for-loop iterate over *all* possible types.

Reply:
I'm fine doing this without touching BasePersistence for now. The behavior in your PR happens to work out nicely for transactional ones which currently just ignore the typeCode. If we do change the behavior somewhere though, or if we add for-loops with different semantics let's make sure we revisit to avoid causing the current transactional impls to do redundant fanout.

Reply:
Yes. That sounds reasonable to me.